File: vhost.php
<?php
/**
 * VirtualHost Generator
 * (For use with Apache 2)
 *
 * @author Jesse Boyer <hello@jream.com>
 * @copyright Copyright (c) 2014, Jese Boyer
 * @license http://opensource.org/licenses/gpl-license.php GNU Public License
 * @version 0.1
 *
 * @usage
 *     Open a Terminal and follow the instructions after typing:
 *     $ php vhost.php
 */
namespace JREAM;

class VirtualHost
{

    protected $address = null;
    protected $server_name = null;
    protected $document_root = null;
    protected $server_path = null;
    protected $output = null;

    /**
     * Inits the Chain
     *
     * @return void
     */
    public function init() {
        $this->setVirtualHostAddress();
    }

    protected function setVirtualHostAddress() {
        echo 'VirtaulHost Address (default=127.0.0.1): ';
        $this->address = trim(fgets(STDIN));
        if (!$this->address) {
            $this->address = '127.0.0.1';
        }

        $this->setServerName();
    }

    protected function setServerName() {
        echo 'VirtualHost ServerName (example: unicorn): ';
        $this->server_name = trim(fgets(STDIN));
        if (!$this->server_name) {
            echo "[Error]: You must provide a server name!\n";
            $this->setServerName();
        }
        $this->setDocumentRoot();
    }

    protected function setDocumentRoot() {
        $default_document_root = "/vagrant/www/{$this->server_name}/public";
        echo "VirtualHost DocumentRoot (default=$default_document_root): ";
        $this->document_root = trim(fgets(STDIN));
        if (!$this->document_root) {
            $this->document_root = $default_document_root;
        }
        $this->setServerPath();
    }

    protected function setServerPath() {
        echo "Use VirtualHost ServerPath? (For Multiple VHosts) (default=y) [y/n]:";
        $this->server_path = trim(fgets(STDIN));
        echo $this->server_path;
        $this->server_path = sprintf("ServerPath /%s", $this->server_name);
        echo $this->server_path;

        if ($this->server_path == 'n') {
            $this->server_path = '';
        }

        echo $this->server_path;
        $this->confirmOutput();
    }

    protected function confirmOutput() {
        $this->output = "
        <VirtualHost {$this->address}>
            ServerName {$this->server_name}
            DocumentRoot {$this->document_root}
            {$this->server_path}
        </VirtualHost>

        <Directory \"{$this->document_root}\">
            Options Indexes FollowSymLinks
            AllowOverride All
            Require all granted
        </Directory>\n
        ";

        echo "\nThe following will be the output of your {$this->server_name}.conf file: \n";
        echo "\n-------------------";

        echo $this->output;

        echo "\nSite URL: {$this->address}/{$this->server_name}";
        echo "\nPublic Directory: " . $this->document_root;

        echo "\nDoes the above look correct? (startover=n) [y/n]:";
        $correct = trim(fgets(STDIN));
        echo $correct;
        if ($correct == 'n') {
            $this->setVirtualHostAddress();
            return;
        }

        $this->confirmWriteFile();
    }

    private function confirmWriteFile() {
        echo "\nWould you like to save the file in the output/ directory? (default=y) [y/n]:";
        $boolean = trim(fgets(STDIN));
        if ($boolean == 'n') {
            return;
        }

        $fp = fopen('output/' . $this->server_name . '.conf', 'w');
        fwrite($fp, $this->output);
        fclose($fp);

        echo "\n\nFile has been saved to {$this->server_name}.conf\n\n";
    }

}


if (!count(debug_backtrace()))
{
    $vhost = new \JREAM\VirtualHost();
    $vhost->init();
}

// End of File
// ---------------------------------------------------------------------

File: vhost.py
"""
@desc       Simple Python Script to create a VHost template.
@author     CodeZeus
@version    0.1
@usage  $ python vhost.py create <servername>
        $ sudo python vhost.py activate <servername>
"""
import os, sys, re
from string import Template

lines = "--------------------------------------"

available_commands = ['create', 'activate']

# ---------------------------------------------------------------------
# Default error message
# ---------------------------------------------------------------------
def error():
    print "\nInvalid command: Please provide:"
    for cmd in available_commands:
        print "\t{0} <servername>".format(cmd)
    sys.exit()

# ---------------------------------------------------------------------
# Prepare the commands
# ---------------------------------------------------------------------
if len(sys.argv) < 3:
    error()

command = sys.argv[1]
param   = sys.argv[2]


if command not in available_commands:
    error()

# ---------------------------------------------------------------------
# Creates an apache vhost template
# ---------------------------------------------------------------------
def create(servername):

    print "This will create an Apache2 VirtualHost template."
    print lines

    if not re.match("^[a-zA-Z]*$", servername):
        print "Error: Only alphabetical characters are allowed."
        sys.exit()

    if len(servername) > 20:
        print "Error: Your servername can only be up to 20 characters."
        sys.exit()

    tpl = Template("""<VirtualHost 127.0.0.1>
        ServerName $servername
        DocumentRoot /vagrant/www/$servername/public
        ServerPath /$servername
    </VirtualHost>

    <Directory "/vagrant/www/$servername/public">
        Options Indexes FollowSymLinks
        AllowOverride All
        Require all granted
    </Directory>
    """)

    output = tpl.safe_substitute({'servername': servername})
    print '\n-- {0}.conf --'.format(servername)
    print output

    confirm = raw_input("Create the template? (You can modify it once created) [Y/n]?: ")
    if (confirm is not 'Y'):
        print "Cancelling."
        sys.exit()

    filename = servername + ".conf"
    conf = open(filename, "w")
    conf.write(output)
    conf.close()


    print "\nFile Created: {0}".format(filename)
    print "To activate your VHost run: "
    print "$ sudo python vhost.py activate {0}".format(servername)

# ---------------------------------------------------------------------
# Activates an Apache VHost
# ---------------------------------------------------------------------
def activate(servername):
    if os.getuid() != 0:
        print("$ sudo is required.")
        sys.exit()

    os.system('sudo mv {0}.conf /etc/apache2/sites-available'.format(servername))
    os.system('sudo a2ensite {0}'.format(servername))
    os.system('sudo service apache2 reload')
    os.system('cd /vagrant/www')
    print "To create a new Phalcon project: $ sudo phalcon project {0}".format(servername)
    print "Otherwise "

# ---------------------------------------------------------------------
# Call a method
# ---------------------------------------------------------------------
locals()[command](param)

# End of File
# ---------------------------------------------------------------------